File: lintcode/naive/binary_tree_maximum_node/cpp/binary_tree_maximum_node.cpp
#include <cmath>

class Solution {
public:
    /**
     * @param root the root of binary tree
     * @return the max node
     */
    TreeNode* maxNode(TreeNode *root) 
    {
        if (root == nullptr) {
            return nullptr;
        }
        
        auto curmax = root;
        
        maxNodeHelper(root, curmax);
        
        return curmax;
    }
    
private:
    void maxNodeHelper(TreeNode *root, TreeNode *&curmax)
    {
        if (root == nullptr) {
            return;
        }
        
        curmax = std::max(root, curmax, 
            [] (const TreeNode *left, const TreeNode *right) {
                return left->val < right->val;
            });
        
        maxNodeHelper(root->left,  curmax);
        maxNodeHelper(root->right, curmax);
    }
};

File: lintcode/naive/binary_tree_maximum_node/py/binary_tree_maximum_node.py
class Solution:
    # @param {TreeNode} root the root of binary tree
    # @return {TreeNode} the max node 
    def maxNode(self, root):
        def helper(root, curmax):
            if root == None:
                return curmax
                
            curmax = max([root, curmax], key=lambda x: x.val)
            
            return max(
                curmax, 
                helper(root.left, curmax), 
                helper(root.right, curmax),
                key=lambda x: x.val)
            
        if root == None:
            return None
        
        return helper(root, root)
